### DISCUSSION TITLE
r:terraform-aws-lambda aws gateway and lambda connection on different route

### DISCUSSION THREAD

**Discussion by npadwal421** [gruntwork-io/knowledge-base#513] [Help]
I am using gruntwork module for api setup with multiple route to same lambda function.

  route_config = {
    "GET /api/ping/{abc}" = {
      lambda_function_arn = module.lambda_function.function_arn
       payload_format_version = "2.0"
    }
    "GET /swagger/index.html" = {
      lambda_function_arn = module.lambda_function.function_arn
       payload_format_version = "2.0"
    }

issue I am getting is "ResourceConflictException: The statement id". When I gone through the grountwork module lambda-http-api-gateway I found it 

```terraform
resource "aws_lambda_permission" "allow_invoke" {
  for_each = var.route_config

  statement_id  = "AllowAPIGateway${var.name}"
  action        = "lambda:InvokeFunction"
  function_name = each.value.lambda_function_arn
  principal     = "apigateway.amazonaws.com"
  source_arn    = "${aws_apigatewayv2_api.api.execution_arn}/*/${local.route_key_split[each.key].method}${local.route_key_split[each.key].path}"
}
```

statement id we cant have multiple which make it conflict. Need help on this.
  

---

<ins datetime="2022-07-20T21:08:31Z">
  <p><a href="https://support.gruntwork.io/hc/requests/109025">Tracked in ticket #109025</a></p>
</ins>

**Comment by Etiene** [comment]
Hey, thanks for bringing this to our attention, I'm gonna try to reproduce this error and get back to you in a bit.

**Comment by Etiene** [comment]
Hi, @npadwal421. The purpose of these routes in the `lambda-http-api-gateway` module is to send different types of requests to different functions. You can see a full example of that here: https://github.com/gruntwork-io/terraform-aws-lambda/blob/master/examples/lambda-http-api-gateway/main.tf

But if you're trying to forward diferrent requests to the same function, and handle it internally at the function, perhaps you can use the 
[`api-gateway-proxy` module](https://github.com/gruntwork-io/terraform-aws-lambda/tree/master/modules/api-gateway-proxy) and use a catch-all configuration. 

Here's an example: https://github.com/gruntwork-io/terraform-aws-lambda/tree/master/examples/lambda-service/path-routing-with-catchall

**Comment by npadwal421** [comment]
@Etiene : thank you for your replay but the example which you send is doing call for 3 different lambda function.
We just need to fix one think in  " lambda-http-api-gateway " module is  ```
resource "aws_lambda_permission" "allow_invoke" {
  for_each = var.route_config
  statement_id  = "AllowAPIGateway${var.name}"
```
**"AllowAPIGateway${var.name}"**   var.name need to be list

**Reply by npadwal421** [reply]
Basically I am trying http api gateway ----- lambda on different route

**Reply by npadwal421** [reply]
@Etiene I tested with api-gateway-proxy. module as well as you suggested that gave me same error```
 module "api_gateway" {
  # When using these modules in your own templates, you will need to use a Git URL with a ref attribute that pins you
  # to a specific version of the modules, such as the following example:
  # source = "git::git@github.com:gruntwork-io/terraform-aws-lambda.git//modules/api-gateway-proxy?ref=v1.0.8"
  source = "git::ssh://git@opslab-i3-git.intuitivecloud.net:2222/isi-gruntwork/terraform-aws-lambda.git//modules/api-gateway-proxy?ref=v0.19.3"
  providers = {
    aws = aws
    # NOTE: this is only necessary if you are configuring an EDGE domain. For other cases, this can be configured to the
    # base `aws` provider.
    aws.us_east_1 = aws.us_east_1
  }
  api_name = var.name
  lambda_functions = {
    # Sets up the following path based routing:
    # - hello/* => hello_lambda function (only returns "hello")
    # - world/* => world_lambda function (only returns "world")
    # - everything else => hello_world_lambda function (only returns "hello world")
    "api"    = "fleet-restapi-test"
    "swagger"= "fleet-restapi-test"
  }
}
```


Error 
```
Error: Error adding new Lambda Permission for fleet-restapi-test: ResourceConflictException: The statement id (AllowExecutionFromApiGateway) provided already exists. Please provide a new statement id, or remove the existing statement.
│ {
│   RespMetadata: {
│     StatusCode: 409,
│     RequestID: "520e057a-18a9-456b-8d5e-d62253072353"
│   },
│   Message_: "The statement id (AllowExecutionFromApiGateway) provided already exists. Please provide a new statement id, or remove the existing statement.",
│   Type: "User"
│ }
```

**Reply by Etiene** [reply]
I could create an issue in our repo as an enhancement feature request for using the same function with different routes apart from the catch all case. 

But in the meantime, I have some questions, the configuration below doesn't suit your needs? This is a "catch all" configuration and should work both for `/api` and `/swagger`. How are you handling the different behaviors internally in the function? Do you intend on adding more routes? 
```
lambda_functions = {
    ""      = "fleet-restapi-test"
}
```
And if this doesn't work, is deploying the same function code twice, the way it is done in [this example](https://github.com/gruntwork-io/terraform-aws-lambda/blob/master/examples/lambda-http-api-gateway/main.tf), an option? I believe AWS lambda only inccurs costs when it executes, but this page should have more info about it: https://aws.amazon.com/lambda/pricing/

**Reply by npadwal421** [reply]
@Etiene  I am using Lambda as a container and that needed to be in VPC to connect to DB, for each route if I cannot deploy 1 lambda.
for time being I am using " https://github.com/terraform-aws-modules/terraform-aws-apigateway-v2 "  module where they are not event specifying statement id and it just work fine

**Reply by Etiene** [reply]
Hi @npadwal421, sorry, I didn't understand what you meant here:
>that needed to be in VPC to connect to DB, for each route if I cannot deploy 1 lambda.

I also still didn't understand why deploying the lambda twice or using a generic catch all route would not work for you. Could you please elaborate a bit more?

**Reply by npadwal421** [reply]
actually for just 1 application 4 Environment DEV, UAT , STAGE, PROD 4 path means 4 lambda each  4 * 4 = 16 lambda, diffult to manage because we have 10 more applications are in pipeline

**Reply by Etiene** [reply]
So you have 4 paths per API gateway, and ideally you want all of them to go the same lambda function?
